### ISSUE TITLE
hlint makes suggestions that break for esqueleto's `select $ from $`

### ISSUE THREAD

**Issue by wolftune** [ndmitchell/hlint#140]
As shown in examples throughout [esqueleto](https://hackage.haskell.org/package/esqueleto-2.2.3/docs/Database-Esqueleto.html), `select $ from $` is an important and valid syntax. hlint consistently warns to remove extraneous `$` but that warning is wrong and breaks compiling.

**Comment by ndmitchell** [comment]
Can I have a small example? I tried the first one I came across in the docs:

    main :: IO ()
    main =
        select $
        from $ \p -> do
        where_ (p ^. PersonName ==. val "John")
        return p

Hlint doesn't give any suggestions there, which seems correct.

**Comment by wolftune** [comment]
Apologies, it looks like I misunderstood. There's actually only a minor issue. When I read "Redundant $", I too it to be comparable to saying that one of the `$` bits was completely unnecessary. However, upon closer inspection of the suggested fix, it is actually a suggestion to remove one of them by changing the scope of a parens pair.

When I simply removed the `$` that I saw gone in the suggestion, things broke. But when I fixed the parens, it worked. So it isn't really wrong. If anything at all, it seems maybe "Redundant $" could be changed to something else like "Adjust enclosures to eliminate extra $" — that wording fits better the other thread where I suggested changing terminology from error/warning to warning/suggestion.

Anyway, closing, but here's a case from our code for reference:

```Haskell
Model/User/Internal.hs:144:7: Warning: Redundant $
Found:
  (\case
       [] -> Nothing
       (x : _) -> unValue x)
    <$>
    (select $
       from $
         \ user ->
           do where_ $ user ^. UserId ==. val user_id
              return $ user ^. UserEmail)
Why not:
  (\case
       [] -> Nothing
       (x : _) -> unValue x)
    <$>
    select
      (from $
         \ user ->
           do where_ $ user ^. UserId ==. val user_id
              return $ user ^. UserEmail)
```

**Comment by wolftune** [comment]
Incidentally, on further though, the "redundant $" message and the solution may be not ideal, as maybe the ideal is to keep the `select $ from $` syntax and there's extraneous parens actually…

**Comment by ndmitchell** [comment]
I've improved the message so in your case it now says "Warning: Move brackets to avoid $" - which I think is more accurate. I agree that it seems a fundamentally different type of hint.

Now that it gets a different name you can ignore it specifically. While the hint is true, and generally a good thing to do, I appreciate that for esqueleto the `select $ from $` idiom is a standard pattern, so making it different is probably a bad thing to do.

**Comment by wolftune** [comment]
heh, now you tell me after I went and made the switch… we'll see what others say on the project; the wording fix is great, thanks!
